### ISSUE TITLE
n_estimators of best model is really large (32768)

### ISSUE THREAD

**Issue by stepthom** [microsoft/FLAML#65]
After running FLAML for several hours, I noticed in the log that the best model was xgboost with n_estimators set to 32768:

```
{"record_id": 24, "iter_per_learner": 55, "logged_metric": false, "trial_time": 1509.8762745857239, "total_search_time": 41334.96948957443, "validation_loss": 0.25501297475819773, "config": {"n_estimators": 32768.0, "max_leaves": 186.0, "min_child_weight": 0.22536063808245474, "learning_rate": 0.05398963108662436,       "subsample": 0.9173715591862044, "colsample_bylevel": 0.9005345477364418, "colsample_bytree": 0.6104797018735161, "reg_alpha": 0.0009765625, "reg_lambda": 1.    92166667176985, "FLAML_sample_size": 55408}, "best_validation_loss": 0.25501297475819773, "best_config": {"n_estimators": 32768.0, "max_leaves": 186.0,          "min_child_weight": 0.22536063808245474, "learning_rate": 0.05398963108662436, "subsample": 0.9173715591862044, "colsample_bylevel": 0.9005345477364418,         "colsample_bytree": 0.6104797018735161, "reg_alpha": 0.0009765625, "reg_lambda": 1.92166667176985, "FLAML_sample_size": 55408}, "learner": "xgboost",            "sample_size": 55408}
{"curr_best_record_id": 24}
```

But that seems excessively large to me and will surely result in an overfit model. (Indeed, the model achieves 99.9% F1 score on the training data set, but only about 75% on a held-out test data set.)

I see in the code that 32768 is set as the upper bound for `n_estimators`:

https://github.com/microsoft/FLAML/blob/06045703bf90f58f3b89acb01b50c8fe2c65b93b/flaml/model.py#L307-L313

I'm just wondering if this upper bound is intentionally set so high, or if this is an oversight.

**Comment by sonichi** [comment]
@stepthom though the upper bound is set high, usually it's not chosen in the final model because the overfitting to training data could be detected by checking validation error. What `eval_method` did you use and how large is your training data?

**Comment by qingyun-wu** [comment]
Thank you for sharing your observation. 
An upper bound of n_estimators up to 2^15 = 32768 is indeed intentionally set. It is set high because we want FLAML to have robust performance for various scales of datasets (large datasets do need models with a larger number of estimators). 

As for the performance, as commented by Chi, FLAML evaluates a config by its validation error. So the gap between the training error and validation error does not necessarily affect FLAML’s performance.

**Comment by stepthom** [comment]
Thanks, @sonichi  and @qingyun-wu.

I am pleased with the results (84% F1 score on my test data) but the model has likely overfit. Are there any tips to help avoid overfitting?

Here are some more details:

I am using the default `eval-method`. 

I have split my dataset into three groups. Training has 63K instances, validation has 7K, and test has 8K. All have 219 features.

I have passed the training and validation to FLAML's `fit()`. I noticed that the final model achieves > 99% F1 on train data, 84% F1 on validation, and 84% F1 on test.

```
    "flaml": {
        "automl_settings": {
            "time_budget": 10000,
            "metric": "f1",
            "task": "classififcation",
            "log_file_name": "out/flaml-490df7f7-f408-408c-bd69-cb08635d05f3.log",
            "n_jobs": 10,
            "estimator_list": [
                "lgbm",
                "xgboost",
                "catboost",
                "extra_tree"
            ],
            "model_history": true
        },
        "starttime": "2021-04-16 17:17:57.725633",
        "endtime": "2021-04-16 20:04:39.428663",
        "best_estimator": "lgbm",
        "best_config": {
            "n_estimators": 32768.0,
            "num_leaves": 256.0,
            "min_child_samples": 61.0,
            "learning_rate": 0.18256185548905027,
            "subsample": 0.5931036157175654,
            "log_max_bin": 9.0,
            "colsample_bytree": 0.210049024349,
            "reg_alpha": 0.009329557337506312,
            "reg_lambda": 0.0026196422111234145,
            "FLAML_sample_size": 63201
        },
        "best_f1_val": 0.7414555689585847,
        "best_model": "LGBMClassifier(colsample_bytree=0.210049024349,\n               learning_rate=0.18256185548905027, max_bin=511,\n               min_child_samples=61, n_estimators=32768, n_jobs=10,\n               num_leaves=256, objective='regression',\n               reg_alpha=0.   009329557337506312, reg_lambda=0.0026196422111234145,\n               subsample=0.5931036157175654)",
        "metrics_train": {
            "accuracy": 0.9990506479327859,
            "f1": 0.9983452103457824,
            "recall": 0.9978435209663918,
            "precision": 0.9988485963705097,
            "report": {
                "False": {
                    "precision": 0.9991578463835245,
                    "recall": 0.9996935981156284,
                    "f1-score": 0.9994256504508645,
                    "support": 52219
                },
                "True": {
                    "precision": 0.998539346357495,
                    "recall": 0.9959934438171554,
                    "f1-score": 0.9972647702407001,
                    "support": 10982
                },
                "accuracy": 0.9990506479327859,
                "macro avg": {
                    "precision": 0.9988485963705097,
                    "recall": 0.9978435209663918,
                    "f1-score": 0.9983452103457824,
                    "support": 63201
                },
                "weighted avg": {
                    "precision": 0.9990503739181227,
                    "recall": 0.9990506479327859,
                    "f1-score": 0.9990501692643639,
                    "support": 63201
                }
            }
        },
        "metrics_val": {
            "accuracy": 0.9179837676206749,
            "f1": 0.8469234290198131,
            "recall": 0.8184397945439437,
            "precision": 0.8865312822449526,
            "report": {
                "False": {
                    "precision": 0.9302633758489316,
                    "recall": 0.9731415699185583,
                    "f1-score": 0.951219512195122,
                    "support": 5771
                },
                "True": {
                    "precision": 0.8427991886409736,
                    "recall": 0.6637380191693291,
                    "f1-score": 0.7426273458445041,
                    "support": 1252
                },
                "accuracy": 0.9179837676206749,
                "macro avg": {
                    "precision": 0.8865312822449526,
                    "recall": 0.8184397945439437,
                    "f1-score": 0.8469234290198131,
                    "support": 7023
                },
                "weighted avg": {
                    "precision": 0.914671013271064,
                    "recall": 0.9179837676206749,
                    "f1-score": 0.914033495924159,
                    "support": 7023
                }
            }
        },
        "metrics_test": {
            "accuracy": 0.9175957964885301,
            "f1": 0.8462213434395789,
            "recall": 0.8159253608339273,
            "precision": 0.8894827358007171,
            "report": {
                "False": {
                    "precision": 0.9284119660663789,
                    "recall": 0.9746875,
                    "f1-score": 0.9509871179205731,
                    "support": 6400
                },
                "True": {
                    "precision": 0.8505535055350554,
                    "recall": 0.6571632216678546,
                    "f1-score": 0.7414555689585847,
                    "support": 1403
                },
                "accuracy": 0.9175957964885301,
                "macro avg": {
                    "precision": 0.8894827358007171,
                    "recall": 0.8159253608339273,
                    "f1-score": 0.8462213434395789,
                    "support": 7803
                },
                "weighted avg": {
                    "precision": 0.9144128093157129,
                    "recall": 0.9175957964885301,
                    "f1-score": 0.9133127922517702,
                    "support": 7803
                }
            }
        },

```

**Comment by qingyun-wu** [comment]
Hi @stepthom ,

One possible way to avoid overfitting is to use an evaluation metric that discounts the training loss in AutomL.  You can set a customized metric in FLAML through the 'metric' argument: https://github.com/microsoft/FLAML/blob/06045703bf90f58f3b89acb01b50c8fe2c65b93b/flaml/automl.py#L811

You can find an example of such a customized metric here: https://github.com/microsoft/FLAML/blob/06045703bf90f58f3b89acb01b50c8fe2c65b93b/test/test_automl.py#L92

Let us know if you need more help!

**Comment by stepthom** [comment]
Thank you - great suggestion. I am trying out a custom f1 metric now. 👍

**Comment by knoam** [comment]
Is there a way to force an upper bound on n_estimators? Even though my model is not overfitting, it is still impractically slow to train since my dataset is large.

**Comment by sonichi** [comment]
> Is there a way to force an upper bound on n_estimators? Even though my model is not overfitting, it is still impractically slow to train since my dataset is large.

@knoam There are multiple ways to solve your problem.
1. You can set `train_time_limit` in AutoML.fit() to restrict the training time per model, without explicitly modifying the upper bound of n_estimators.
2. You can inherit the default estimator and override its [search space](https://microsoft.github.io/FLAML/docs/Use-Cases/Task-Oriented-AutoML#search-space) to use a smaller upper bound.

Please let me know if this solves your problem. Also, I'd appreciate your feedback in this survey: https://docs.google.com/forms/d/e/1FAIpQLSd0u4BbSzr-m9hgM8rmPlYB_JJoFXeHhuoXI9YFbHDRWrDlOQ/viewform?usp=sf_link
